### PR TITLE
[AIP-192] Add rule for leading comments.

### LINKED ISSUE #174: AIP-192: Documentation / leading & trailing comments
Old linter rule: https://g3doc.corp.google.com/google/api/tools/linter/g3doc/rules/leading-and-trailing-comment.md?cl=head

### PR DESCRIPTION
**Pull request by lukesneeringer**
Also adds a `DescriptorRule` that pokes everything.
I will migrate the AIP-140 abbreviations component to that
after this is merged.

Fixes #174.

### CODE REVIEW

File: rules/aip0192/aip0192.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+		for len(c) > 0 {
+			// Anything before the `(--` is external string.
+			open := strings.SplitN(c, "(--", 2)
+			if ex := strings.TrimSpace(open[0]); ex != "" {
+				answer.External = append(answer.External, ex)
+			}
+			if len(open) > 1 {
+				c = strings.TrimSpace(open[1])
+			} else {
+				c = ""
```
Comment (mingzhi) [inline rules/aip0192/aip0192.go]: I think we can break here.

File: rules/aip0192/aip0192.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+			// Now that the opening component is tokenized, anything before
+			// the `--)` is internal string.
+			close := strings.SplitN(c, "--)", 2)
+			if in := strings.TrimSpace(close[0]); in != "" {
+				answer.Internal = append(answer.Internal, in)
+			}
+			if len(close) > 1 {
+				c = strings.TrimSpace(close[1])
+			} else {
+				c = ""
```
Comment (mingzhi) [inline rules/aip0192/aip0192.go]: We can break here.

File: rules/aip0192/leading_comments_test.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+func TestOnlyLeadingComments(t *testing.T) {
+	tests := []struct {
+		testName string
+		Detached string
+		Trailing string
+		problems testutils.Problems
+	}{
+		{"ValidNone", "", "", testutils.Problems{}},
+		{"ValidInternalDetached", "// (-- detached --)", "", testutils.Problems{}},
+		{"ValidInternalTrailing", "// (-- trailing --)", "", testutils.Problems{}},
```
Comment (mingzhi) [inline rules/aip0192/leading_comments_test.go]: This is still "detached" comments, not the trailing one.

File: rules/aip0192/leading_comments_test.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+		{"ValidInternalTrailingNoClose", "// (-- trailing", "", testutils.Problems{}},
```
Comment (mingzhi) [inline rules/aip0192/leading_comments_test.go]: Same here, it is a "detached" comment, not the trailing one you wanted.

File: lint/rule.go [hunk 5]
```diff
@@ -251,21 +251,70 @@ func (r *EnumRule) GetURI() string {
...
+	for _, enum := range getAllEnums(fd) {
 		if r.OnlyIf == nil || r.OnlyIf(enum) {
 			problems = append(problems, r.LintEnum(enum)...)
 		}
 	}
+	return problems
+}
+
+// DescriptorRule defines a lint rule that is run on every descriptor
+// in the file (but not the file itself).
```
[the remark this answers is not in the document]
Comment (lukesneeringer) [reply lint/rule.go]: Done.

File: rules/aip0192/aip0192.go (hunk 2 above)
Comment (lukesneeringer) [reply rules/aip0192/aip0192.go]: Done.

File: rules/aip0192/aip0192.go (hunk 1 above)
Comment (lukesneeringer) [reply rules/aip0192/aip0192.go]: Done.

File: rules/aip0192/leading_comments_test.go (hunk 3 above)
Comment (lukesneeringer) [reply rules/aip0192/leading_comments_test.go]: Done.

File: rules/aip0192/leading_comments_test.go (hunk 4 above)
Comment (lukesneeringer) [reply rules/aip0192/leading_comments_test.go]: Done.

**Pull request merged**